### PR TITLE
Stack and functions

### PR DESCRIPTION
**Pull request by JKTKops**
Brings in the stack & functions extension as well as some minor fixes I had locally.

Additionally brings in several new features from Mega which were intended to go straight to master:
  - `.set SYMBOL VALUE` directive, for creating named constants
  - `.half/.word/.dword`, for including literal data in the binary
  - support for character literals
  - bugfixes related to the previous pass system's reset mechanism

### CODE REVIEW

File: etca_asm/core.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
 from __future__ import annotations
 
+from ast import literal_eval
+from cgitb import enable
```
Comment (MegaIng) [inline etca_asm/core.py]: I assume this import was an accident?

File: etca_asm/extensions/stack_and_functions.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+from wsgiref import validate
```
Comment (MegaIng) [inline etca_asm/extensions/stack_and_functions.py]: And this import is probably also accidental?

File: etca_asm/extensions/stack_and_functions.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+from etca_asm.core import Extension, reject
+from etca_asm.base_isa import CONDITION_NAMES, validate_registers, build
+
+functions = Extension(2, "functions", "Stack and Functions")
+Register = tuple[int | None, int]
+
+def oneof(*names):
```
Comment (MegaIng) [inline etca_asm/extensions/stack_and_functions.py]: This function can now be imported from .core

**Review by MegaIng — CHANGES_REQUESTED** [review]

**Review by MegaIng — APPROVED** [review]

File: etca_asm/extensions/stack_and_functions.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+from etca_asm.core import Extension, reject, oneof
+from etca_asm.base_isa import CONDITION_NAMES, validate_registers, build
+
+functions = Extension(2, "functions", "Stack and Functions")
+Register = tuple[int | None, int]
+
+REGISTERS = {
```
Comment (EnderShadow) [inline etca_asm/extensions/stack_and_functions.py:7]: These don't match the spec. v0 should be a2 and a0 and a1 should also be v0 and v1.

File: etca_asm/extensions/stack_and_functions.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    return build((0b00,2), (cxt.register_sizes[size],2), (0xD,4), (6,3), (src,3), (0b00,2))
+
+@functions.inst(f'"push" size_postfix immediate')
+def push_register_imm(cxt, inst_size, imm: int):
+    size, () = validate_registers(cxt, inst_size=inst_size)
+    reject(
+        not isinstance(imm, int) or not (0 <= imm < 32),
+        f"Invalidate immediate {imm} for op `push'"
+    )
+    return build((0b01,2), (cxt.register_sizes[size],2), (0xD,4), (0b000,3), (imm,5))
```
Comment (EnderShadow) [inline etca_asm/extensions/stack_and_functions.py]: That should be `(6, 3)`

**Review by EnderShadow — CHANGES_REQUESTED** [review]

File: etca_asm/extensions/stack_and_functions.py (hunk 5 above)
Comment (JKTKops) [reply etca_asm/extensions/stack_and_functions.py]: ree I thought I fixed this one

File: etca_asm/core.py [hunk 6]
```diff
@@ -98,64 +105,98 @@ def core_init(context):
...
 
+def oneof(*names):
+    names = sorted(names, key=len, reverse=True)
+    return f"({'|'.join(names)})"
+
+
+_WORD_SIZES = {
+    '.half': 1,
+    '.word': 2,
+    '.dword': 4,
```
Comment (EnderShadow) [inline etca_asm/core.py:128]: Should we add .qword to this just so we don't forget?

Comment (EnderShadow) [reply etca_asm/core.py:128]: And maybe something for arrays/strings?

Comment (JKTKops) [reply etca_asm/core.py:128]: Let's do that separately, this PR is already fairly scatterbrained.

**Review by EnderShadow — APPROVED** [review]

**Pull request merged**